// @11labs/react/dist/lib.cjs
var n = require("react"),
  e = require("@elevenlabs/client");
function r() {
  return (
    (r = Object.assign
      ? Object.assign.bind()
      : function (n) {
          for (var e = 1; e < arguments.length; e++) {
            var r = arguments[e];
            for (var t in r) ({}).hasOwnProperty.call(r, t) && (n[t] = r[t]);
          }
          return n;
        }),
    r.apply(null, arguments)
  );
}
var t = ["micMuted", "volume", "serverLocation"];
function u(n) {
  switch ((void 0 === n && (n = "us"), n)) {
    case "eu-residency":
    case "in-residency":
    case "us":
    case "global":
      return n;
    default:
      return (
        console.warn(
          "[ConversationalAI] Invalid server-location: " +
            n +
            '. Defaulting to "us"'
        ),
        "us"
      );
  }
}
function l(n) {
  return {
    us: "wss://api.elevenlabs.io",
    "eu-residency": "wss://api.eu.residency.elevenlabs.io",
    "in-residency": "wss://api.in.residency.elevenlabs.io",
    global: "wss://api.elevenlabs.io",
  }[n];
}
(Object.defineProperty(exports, "postOverallFeedback", {
  enumerable: !0,
  get: function () {
    return e.postOverallFeedback;
  },
}),
  (exports.getOriginForLocation = l),
  (exports.parseLocation = u),
  (exports.useConversation = function (o) {
    void 0 === o && (o = {});
    var i = o.micMuted,
      s = o.volume,
      c = o.serverLocation,
      a = (function (n, e) {
        if (null == n) return {};
        var r = {};
        for (var t in n)
          if ({}.hasOwnProperty.call(n, t)) {
            if (-1 !== e.indexOf(t)) continue;
            r[t] = n[t];
          }
        return r;
      })(o, t),
      d = n.useRef(null),
      v = n.useRef(null),
      f = n.useState("disconnected"),
      g = f[0],
      p = f[1],
      m = n.useState(!1),
      b = m[0],
      y = m[1],
      h = n.useState("listening"),
      C = h[0],
      M = h[1];
    return (
      n.useEffect(
        function () {
          var n;
          void 0 !== i &&
            (null == d || null == (n = d.current) || n.setMicMuted(i));
        },
        [i]
      ),
      n.useEffect(
        function () {
          var n;
          void 0 !== s &&
            (null == d ||
              null == (n = d.current) ||
              n.setVolume({ volume: s }));
        },
        [s]
      ),
      n.useEffect(function () {
        return function () {
          var n;
          null == (n = d.current) || n.endSession();
        };
      }, []),
      {
        startSession: function (n) {
          try {
            var t,
              o,
              f = function (t) {
                return o
                  ? t
                  : (function (t, o) {
                      try {
                        var f =
                          ((I = l(
                            u((null == n ? void 0 : n.serverLocation) || c)
                          )),
                          (v.current = e.Conversation.startSession(
                            r({}, null != a ? a : {}, null != n ? n : {}, {
                              origin: I,
                              overrides: r(
                                {},
                                null != (g = null == a ? void 0 : a.overrides)
                                  ? g
                                  : {},
                                null != (m = null == n ? void 0 : n.overrides)
                                  ? m
                                  : {},
                                {
                                  client: r(
                                    {},
                                    null !=
                                      (b =
                                        null == a || null == (h = a.overrides)
                                          ? void 0
                                          : h.client)
                                      ? b
                                      : {},
                                    null !=
                                      (C =
                                        null == n || null == (O = n.overrides)
                                          ? void 0
                                          : O.client)
                                      ? C
                                      : {},
                                    {
                                      source:
                                        (null == n ||
                                        null == (S = n.overrides) ||
                                        null == (S = S.client)
                                          ? void 0
                                          : S.source) ||
                                        (null == a ||
                                        null == (F = a.overrides) ||
                                        null == (F = F.client)
                                          ? void 0
                                          : F.source) ||
                                        "react_sdk",
                                      version:
                                        (null == n ||
                                        null == (P = n.overrides) ||
                                        null == (P = P.client)
                                          ? void 0
                                          : P.version) ||
                                        (null == a ||
                                        null == (D = a.overrides) ||
                                        null == (D = D.client)
                                          ? void 0
                                          : D.version) ||
                                        "0.4.3",
                                    }
                                  ),
                                }
                              ),
                              onConnect:
                                (null == n ? void 0 : n.onConnect) ||
                                (null == a ? void 0 : a.onConnect),
                              onDisconnect:
                                (null == n ? void 0 : n.onDisconnect) ||
                                (null == a ? void 0 : a.onDisconnect),
                              onError:
                                (null == n ? void 0 : n.onError) ||
                                (null == a ? void 0 : a.onError),
                              onMessage:
                                (null == n ? void 0 : n.onMessage) ||
                                (null == a ? void 0 : a.onMessage),
                              onAudio:
                                (null == n ? void 0 : n.onAudio) ||
                                (null == a ? void 0 : a.onAudio),
                              onDebug:
                                (null == n ? void 0 : n.onDebug) ||
                                (null == a ? void 0 : a.onDebug),
                              onUnhandledClientToolCall:
                                (null == n
                                  ? void 0
                                  : n.onUnhandledClientToolCall) ||
                                (null == a
                                  ? void 0
                                  : a.onUnhandledClientToolCall),
                              onModeChange: function (n) {
                                M(n.mode);
                              },
                              onStatusChange: function (n) {
                                p(n.status);
                              },
                              onCanSendFeedbackChange: function (n) {
                                y(n.canSendFeedback);
                              },
                            })
                          )),
                          Promise.resolve(v.current).then(function (n) {
                            return (
                              (d.current = n),
                              void 0 !== i && d.current.setMicMuted(i),
                              void 0 !== s &&
                                d.current.setVolume({ volume: s }),
                              d.current.getId()
                            );
                          }));
                      } catch (n) {
                        return o(!0, n);
                      }
                      var g, m, b, h, C, O, S, F, P, D, I;
                      return f && f.then
                        ? f.then(o.bind(null, !1), o.bind(null, !0))
                        : o(!1, f);
                    })(0, function (n, e) {
                      if (((v.current = null), n)) throw e;
                      return e;
                    });
              };
            if (null != (t = d.current) && t.isOpen())
              return Promise.resolve(d.current.getId());
            var g = (function () {
              if (v.current)
                return Promise.resolve(v.current).then(function (n) {
                  var e = n.getId();
                  return ((o = 1), e);
                });
            })();
            return Promise.resolve(g && g.then ? g.then(f) : f(g));
          } catch (n) {
            return Promise.reject(n);
          }
        },
        endSession: function () {
          try {
            var n = d.current;
            return (
              (d.current = null),
              Promise.resolve(null == n ? void 0 : n.endSession()).then(
                function () {}
              )
            );
          } catch (n) {
            return Promise.reject(n);
          }
        },
        setVolume: function (n) {
          var e;
          null == (e = d.current) || e.setVolume({ volume: n.volume });
        },
        getInputByteFrequencyData: function () {
          var n;
          return null == (n = d.current)
            ? void 0
            : n.getInputByteFrequencyData();
        },
        getOutputByteFrequencyData: function () {
          var n;
          return null == (n = d.current)
            ? void 0
            : n.getOutputByteFrequencyData();
        },
        getInputVolume: function () {
          var n, e;
          return null !=
            (n = null == (e = d.current) ? void 0 : e.getInputVolume())
            ? n
            : 0;
        },
        getOutputVolume: function () {
          var n, e;
          return null !=
            (n = null == (e = d.current) ? void 0 : e.getOutputVolume())
            ? n
            : 0;
        },
        sendFeedback: function (n) {
          var e;
          null == (e = d.current) || e.sendFeedback(n);
        },
        getId: function () {
          var n;
          return null == (n = d.current) ? void 0 : n.getId();
        },
        sendContextualUpdate: function (n) {
          var e;
          null == (e = d.current) || e.sendContextualUpdate(n);
        },
        sendUserMessage: function (n) {
          var e;
          null == (e = d.current) || e.sendUserMessage(n);
        },
        sendUserActivity: function () {
          var n;
          null == (n = d.current) || n.sendUserActivity();
        },
        sendMCPToolApprovalResult: function (n, e) {
          var r;
          null == (r = d.current) || r.sendMCPToolApprovalResult(n, e);
        },
        status: g,
        canSendFeedback: b,
        micMuted: i,
        isSpeaking: "speaking" === C,
      }
    );
  }));
