// @11labs/react/dist/lib.module.js
import { useRef as n, useState as e, useEffect as r } from "react";
import { Conversation as l } from "@elevenlabs/client";
export { postOverallFeedback } from "@elevenlabs/client";
function u() {
  return (
    (u = Object.assign
      ? Object.assign.bind()
      : function (n) {
          for (var e = 1; e < arguments.length; e++) {
            var r = arguments[e];
            for (var l in r) ({}).hasOwnProperty.call(r, l) && (n[l] = r[l]);
          }
          return n;
        }),
    u.apply(null, arguments)
  );
}
var t = ["micMuted", "volume", "serverLocation"];
function o(n) {
  switch ((void 0 === n && (n = "us"), n)) {
    case "eu-residency":
    case "in-residency":
    case "us":
    case "global":
      return n;
    default:
      return (
        console.warn(
          "[ConversationalAI] Invalid server-location: " +
            n +
            '. Defaulting to "us"'
        ),
        "us"
      );
  }
}
function i(n) {
  return {
    us: "wss://api.elevenlabs.io",
    "eu-residency": "wss://api.eu.residency.elevenlabs.io",
    "in-residency": "wss://api.in.residency.elevenlabs.io",
    global: "wss://api.elevenlabs.io",
  }[n];
}
function c(c) {
  void 0 === c && (c = {});
  var s = c.micMuted,
    a = c.volume,
    d = c.serverLocation,
    v = (function (n, e) {
      if (null == n) return {};
      var r = {};
      for (var l in n)
        if ({}.hasOwnProperty.call(n, l)) {
          if (-1 !== e.indexOf(l)) continue;
          r[l] = n[l];
        }
      return r;
    })(c, t),
    f = n(null),
    g = n(null),
    m = e("disconnected"),
    p = m[0],
    b = m[1],
    y = e(!1),
    h = y[0],
    C = y[1],
    M = e("listening"),
    D = M[0],
    I = M[1];
  return (
    r(
      function () {
        var n;
        void 0 !== s &&
          (null == f || null == (n = f.current) || n.setMicMuted(s));
      },
      [s]
    ),
    r(
      function () {
        var n;
        void 0 !== a &&
          (null == f || null == (n = f.current) || n.setVolume({ volume: a }));
      },
      [a]
    ),
    r(function () {
      return function () {
        var n;
        null == (n = f.current) || n.endSession();
      };
    }, []),
    {
      startSession: function (n) {
        try {
          var e,
            r,
            t = function (e) {
              return r
                ? e
                : (function (e, r) {
                    try {
                      var t =
                        ((S = i(
                          o((null == n ? void 0 : n.serverLocation) || d)
                        )),
                        (g.current = l.startSession(
                          u({}, null != v ? v : {}, null != n ? n : {}, {
                            origin: S,
                            overrides: u(
                              {},
                              null != (c = null == v ? void 0 : v.overrides)
                                ? c
                                : {},
                              null != (m = null == n ? void 0 : n.overrides)
                                ? m
                                : {},
                              {
                                client: u(
                                  {},
                                  null !=
                                    (p =
                                      null == v || null == (y = v.overrides)
                                        ? void 0
                                        : y.client)
                                    ? p
                                    : {},
                                  null !=
                                    (h =
                                      null == n || null == (M = n.overrides)
                                        ? void 0
                                        : M.client)
                                    ? h
                                    : {},
                                  {
                                    source:
                                      (null == n ||
                                      null == (D = n.overrides) ||
                                      null == (D = D.client)
                                        ? void 0
                                        : D.source) ||
                                      (null == v ||
                                      null == (O = v.overrides) ||
                                      null == (O = O.client)
                                        ? void 0
                                        : O.source) ||
                                      "react_sdk",
                                    version:
                                      (null == n ||
                                      null == (P = n.overrides) ||
                                      null == (P = P.client)
                                        ? void 0
                                        : P.version) ||
                                      (null == v ||
                                      null == (F = v.overrides) ||
                                      null == (F = F.client)
                                        ? void 0
                                        : F.version) ||
                                      "0.4.3",
                                  }
                                ),
                              }
                            ),
                            onConnect:
                              (null == n ? void 0 : n.onConnect) ||
                              (null == v ? void 0 : v.onConnect),
                            onDisconnect:
                              (null == n ? void 0 : n.onDisconnect) ||
                              (null == v ? void 0 : v.onDisconnect),
                            onError:
                              (null == n ? void 0 : n.onError) ||
                              (null == v ? void 0 : v.onError),
                            onMessage:
                              (null == n ? void 0 : n.onMessage) ||
                              (null == v ? void 0 : v.onMessage),
                            onAudio:
                              (null == n ? void 0 : n.onAudio) ||
                              (null == v ? void 0 : v.onAudio),
                            onDebug:
                              (null == n ? void 0 : n.onDebug) ||
                              (null == v ? void 0 : v.onDebug),
                            onUnhandledClientToolCall:
                              (null == n
                                ? void 0
                                : n.onUnhandledClientToolCall) ||
                              (null == v
                                ? void 0
                                : v.onUnhandledClientToolCall),
                            onModeChange: function (n) {
                              I(n.mode);
                            },
                            onStatusChange: function (n) {
                              b(n.status);
                            },
                            onCanSendFeedbackChange: function (n) {
                              C(n.canSendFeedback);
                            },
                          })
                        )),
                        Promise.resolve(g.current).then(function (n) {
                          return (
                            (f.current = n),
                            void 0 !== s && f.current.setMicMuted(s),
                            void 0 !== a && f.current.setVolume({ volume: a }),
                            f.current.getId()
                          );
                        }));
                    } catch (n) {
                      return r(!0, n);
                    }
                    var c, m, p, y, h, M, D, O, P, F, S;
                    return t && t.then
                      ? t.then(r.bind(null, !1), r.bind(null, !0))
                      : r(!1, t);
                  })(0, function (n, e) {
                    if (((g.current = null), n)) throw e;
                    return e;
                  });
            };
          if (null != (e = f.current) && e.isOpen())
            return Promise.resolve(f.current.getId());
          var c = (function () {
            if (g.current)
              return Promise.resolve(g.current).then(function (n) {
                var e = n.getId();
                return ((r = 1), e);
              });
          })();
          return Promise.resolve(c && c.then ? c.then(t) : t(c));
        } catch (n) {
          return Promise.reject(n);
        }
      },
      endSession: function () {
        try {
          var n = f.current;
          return (
            (f.current = null),
            Promise.resolve(null == n ? void 0 : n.endSession()).then(
              function () {}
            )
          );
        } catch (n) {
          return Promise.reject(n);
        }
      },
      setVolume: function (n) {
        var e;
        null == (e = f.current) || e.setVolume({ volume: n.volume });
      },
      getInputByteFrequencyData: function () {
        var n;
        return null == (n = f.current) ? void 0 : n.getInputByteFrequencyData();
      },
      getOutputByteFrequencyData: function () {
        var n;
        return null == (n = f.current)
          ? void 0
          : n.getOutputByteFrequencyData();
      },
      getInputVolume: function () {
        var n, e;
        return null !=
          (n = null == (e = f.current) ? void 0 : e.getInputVolume())
          ? n
          : 0;
      },
      getOutputVolume: function () {
        var n, e;
        return null !=
          (n = null == (e = f.current) ? void 0 : e.getOutputVolume())
          ? n
          : 0;
      },
      sendFeedback: function (n) {
        var e;
        null == (e = f.current) || e.sendFeedback(n);
      },
      getId: function () {
        var n;
        return null == (n = f.current) ? void 0 : n.getId();
      },
      sendContextualUpdate: function (n) {
        var e;
        null == (e = f.current) || e.sendContextualUpdate(n);
      },
      sendUserMessage: function (n) {
        var e;
        null == (e = f.current) || e.sendUserMessage(n);
      },
      sendUserActivity: function () {
        var n;
        null == (n = f.current) || n.sendUserActivity();
      },
      sendMCPToolApprovalResult: function (n, e) {
        var r;
        null == (r = f.current) || r.sendMCPToolApprovalResult(n, e);
      },
      status: p,
      canSendFeedback: h,
      micMuted: s,
      isSpeaking: "speaking" === D,
    }
  );
}
export { i as getOriginForLocation, o as parseLocation, c as useConversation };
