// @11labs/react/dist/lib.umd.js
!(function (e, n) {
  "object" == typeof exports && "undefined" != typeof module
    ? n(exports, require("react"), require("@elevenlabs/client"))
    : "function" == typeof define && define.amd
      ? define(["exports", "react", "@elevenlabs/client"], n)
      : n(((e || self).react = {}), e.react, e.client);
})(this, function (e, n, r) {
  function t() {
    return (
      (t = Object.assign
        ? Object.assign.bind()
        : function (e) {
            for (var n = 1; n < arguments.length; n++) {
              var r = arguments[n];
              for (var t in r) ({}).hasOwnProperty.call(r, t) && (e[t] = r[t]);
            }
            return e;
          }),
      t.apply(null, arguments)
    );
  }
  var u = ["micMuted", "volume", "serverLocation"];
  function l(e) {
    switch ((void 0 === e && (e = "us"), e)) {
      case "eu-residency":
      case "in-residency":
      case "us":
      case "global":
        return e;
      default:
        return (
          console.warn(
            "[ConversationalAI] Invalid server-location: " +
              e +
              '. Defaulting to "us"'
          ),
          "us"
        );
    }
  }
  function o(e) {
    return {
      us: "wss://api.elevenlabs.io",
      "eu-residency": "wss://api.eu.residency.elevenlabs.io",
      "in-residency": "wss://api.in.residency.elevenlabs.io",
      global: "wss://api.elevenlabs.io",
    }[e];
  }
  (Object.defineProperty(e, "postOverallFeedback", {
    enumerable: !0,
    get: function () {
      return r.postOverallFeedback;
    },
  }),
    (e.getOriginForLocation = o),
    (e.parseLocation = l),
    (e.useConversation = function (e) {
      void 0 === e && (e = {});
      var i = e.micMuted,
        c = e.volume,
        s = e.serverLocation,
        a = (function (e, n) {
          if (null == e) return {};
          var r = {};
          for (var t in e)
            if ({}.hasOwnProperty.call(e, t)) {
              if (-1 !== n.indexOf(t)) continue;
              r[t] = e[t];
            }
          return r;
        })(e, u),
        d = n.useRef(null),
        v = n.useRef(null),
        f = n.useState("disconnected"),
        g = f[0],
        p = f[1],
        b = n.useState(!1),
        m = b[0],
        y = b[1],
        h = n.useState("listening"),
        C = h[0],
        M = h[1];
      return (
        n.useEffect(
          function () {
            var e;
            void 0 !== i &&
              (null == d || null == (e = d.current) || e.setMicMuted(i));
          },
          [i]
        ),
        n.useEffect(
          function () {
            var e;
            void 0 !== c &&
              (null == d ||
                null == (e = d.current) ||
                e.setVolume({ volume: c }));
          },
          [c]
        ),
        n.useEffect(function () {
          return function () {
            var e;
            null == (e = d.current) || e.endSession();
          };
        }, []),
        {
          startSession: function (e) {
            try {
              var n,
                u,
                f = function (n) {
                  return u
                    ? n
                    : (function (n, u) {
                        try {
                          var f =
                            ((I = o(
                              l((null == e ? void 0 : e.serverLocation) || s)
                            )),
                            (v.current = r.Conversation.startSession(
                              t({}, null != a ? a : {}, null != e ? e : {}, {
                                origin: I,
                                overrides: t(
                                  {},
                                  null != (g = null == a ? void 0 : a.overrides)
                                    ? g
                                    : {},
                                  null != (b = null == e ? void 0 : e.overrides)
                                    ? b
                                    : {},
                                  {
                                    client: t(
                                      {},
                                      null !=
                                        (m =
                                          null == a || null == (h = a.overrides)
                                            ? void 0
                                            : h.client)
                                        ? m
                                        : {},
                                      null !=
                                        (C =
                                          null == e || null == (O = e.overrides)
                                            ? void 0
                                            : O.client)
                                        ? C
                                        : {},
                                      {
                                        source:
                                          (null == e ||
                                          null == (S = e.overrides) ||
                                          null == (S = S.client)
                                            ? void 0
                                            : S.source) ||
                                          (null == a ||
                                          null == (F = a.overrides) ||
                                          null == (F = F.client)
                                            ? void 0
                                            : F.source) ||
                                          "react_sdk",
                                        version:
                                          (null == e ||
                                          null == (P = e.overrides) ||
                                          null == (P = P.client)
                                            ? void 0
                                            : P.version) ||
                                          (null == a ||
                                          null == (D = a.overrides) ||
                                          null == (D = D.client)
                                            ? void 0
                                            : D.version) ||
                                          "0.4.3",
                                      }
                                    ),
                                  }
                                ),
                                onConnect:
                                  (null == e ? void 0 : e.onConnect) ||
                                  (null == a ? void 0 : a.onConnect),
                                onDisconnect:
                                  (null == e ? void 0 : e.onDisconnect) ||
                                  (null == a ? void 0 : a.onDisconnect),
                                onError:
                                  (null == e ? void 0 : e.onError) ||
                                  (null == a ? void 0 : a.onError),
                                onMessage:
                                  (null == e ? void 0 : e.onMessage) ||
                                  (null == a ? void 0 : a.onMessage),
                                onAudio:
                                  (null == e ? void 0 : e.onAudio) ||
                                  (null == a ? void 0 : a.onAudio),
                                onDebug:
                                  (null == e ? void 0 : e.onDebug) ||
                                  (null == a ? void 0 : a.onDebug),
                                onUnhandledClientToolCall:
                                  (null == e
                                    ? void 0
                                    : e.onUnhandledClientToolCall) ||
                                  (null == a
                                    ? void 0
                                    : a.onUnhandledClientToolCall),
                                onModeChange: function (e) {
                                  M(e.mode);
                                },
                                onStatusChange: function (e) {
                                  p(e.status);
                                },
                                onCanSendFeedbackChange: function (e) {
                                  y(e.canSendFeedback);
                                },
                              })
                            )),
                            Promise.resolve(v.current).then(function (e) {
                              return (
                                (d.current = e),
                                void 0 !== i && d.current.setMicMuted(i),
                                void 0 !== c &&
                                  d.current.setVolume({ volume: c }),
                                d.current.getId()
                              );
                            }));
                        } catch (e) {
                          return u(!0, e);
                        }
                        var g, b, m, h, C, O, S, F, P, D, I;
                        return f && f.then
                          ? f.then(u.bind(null, !1), u.bind(null, !0))
                          : u(!1, f);
                      })(0, function (e, n) {
                        if (((v.current = null), e)) throw n;
                        return n;
                      });
                };
              if (null != (n = d.current) && n.isOpen())
                return Promise.resolve(d.current.getId());
              var g = (function () {
                if (v.current)
                  return Promise.resolve(v.current).then(function (e) {
                    var n = e.getId();
                    return ((u = 1), n);
                  });
              })();
              return Promise.resolve(g && g.then ? g.then(f) : f(g));
            } catch (e) {
              return Promise.reject(e);
            }
          },
          endSession: function () {
            try {
              var e = d.current;
              return (
                (d.current = null),
                Promise.resolve(null == e ? void 0 : e.endSession()).then(
                  function () {}
                )
              );
            } catch (e) {
              return Promise.reject(e);
            }
          },
          setVolume: function (e) {
            var n;
            null == (n = d.current) || n.setVolume({ volume: e.volume });
          },
          getInputByteFrequencyData: function () {
            var e;
            return null == (e = d.current)
              ? void 0
              : e.getInputByteFrequencyData();
          },
          getOutputByteFrequencyData: function () {
            var e;
            return null == (e = d.current)
              ? void 0
              : e.getOutputByteFrequencyData();
          },
          getInputVolume: function () {
            var e, n;
            return null !=
              (e = null == (n = d.current) ? void 0 : n.getInputVolume())
              ? e
              : 0;
          },
          getOutputVolume: function () {
            var e, n;
            return null !=
              (e = null == (n = d.current) ? void 0 : n.getOutputVolume())
              ? e
              : 0;
          },
          sendFeedback: function (e) {
            var n;
            null == (n = d.current) || n.sendFeedback(e);
          },
          getId: function () {
            var e;
            return null == (e = d.current) ? void 0 : e.getId();
          },
          sendContextualUpdate: function (e) {
            var n;
            null == (n = d.current) || n.sendContextualUpdate(e);
          },
          sendUserMessage: function (e) {
            var n;
            null == (n = d.current) || n.sendUserMessage(e);
          },
          sendUserActivity: function () {
            var e;
            null == (e = d.current) || e.sendUserActivity();
          },
          sendMCPToolApprovalResult: function (e, n) {
            var r;
            null == (r = d.current) || r.sendMCPToolApprovalResult(e, n);
          },
          status: g,
          canSendFeedback: m,
          micMuted: i,
          isSpeaking: "speaking" === C,
        }
      );
    }));
});
